### PR TITLE
T28780 (PAYG)

### PR DESCRIPTION
**Pull request by GeorgesStavracas**
✓ pick 79ee0802f st-private: Add support for letter-spacing in ClutterText based StWidgets
✓ pick 543b6388d st-entry: Set text-related CSS properties on the internal ClutterText
✓ pick 2f9689795 payg: Initial implementation of the PAYG manager component
	squash ae2b6ee87 payg: Check if proxy was created
	squash 24a54fbea payg: Wait for the PaygManager before showing the screenshield's dialog
	squash 7ae9e98d1 payg: Show periodic notifications with the remaining PAYG time
	squash b16f98190 payg: Show a notification with the remaining PAYG time on startup
	squash 1a41973e5 paygManager: Understand ExpiryTime in terms of CLOCK_BOOTTIME
	squash 355602f11 paygManager: make timeRemainingSecs() public
✓ pick a94c3e0e0 payg: Add a new screen to allow unlocking PAYG machines
	squash e55326ddb payg: Visual changes to the code entry widget to make it stand out
	squash a2e5c06a9 paygUnlockDialog: delegate validation to paygManager
✓ pick 969edcf8c payg: Integrate the unlocking dialog for PAYG in the screen shield
✓ pick 32a956a04 payg: Make sure to initialize the users list after startup has completed
✓ pick d0f1fa33a payg: Integrate the PAYG unlock screen with the login screen
✓ pick 7149ca06a payg: Don't auto-lock out of user sessions based on PAYG status
✓ pick 410af9d59 customerSupport: Manage vendor-customer-support.ini from its own class
✓ pick 9d7344eee payg: Support image-specific overrides for text in the unlock dialog
✓ pick 38809624b payg: Show image-specific instructions for PAYG, if available
✓ pick 51d9d2daa payg: tweak code entry field for longer codes
✓ pick cbcb45027 messageTray: Don't expire a notification if its focus grabber is focused
✓ pick ba5b0ded0 js/ui/payg.js: PayGo UI utilities
	squash 1358d284d payg: Use CLOCK_BOOTTIME for rate limiting
✓ pick 8f018e019 payg: factor out reminder notifications
✓ pick 6e919d8e8 status/payg: new applet for PAYG account details and actions
✓ pick e9f4eef9d status/payg: add menu action to apply a code before PayGo time is expired
	squash 1684d3d6e config: Remove obsolete notifications
✓ pick 5562d7c70 payg: Add success screen
✓ pick cfd90984c eos-payg: Add prefix and sufix characters for code insert
	squash b766e2eca eos-payg: Add prefix and suffix characters for the payg-tray
✓ pick 634c4ac27 eos-payg: Better segregate Pay As You Go classes
✓ pick 4d4ea2764 eos-payg: Add the character enforcement for payg-tray
✓ pick 4bf57d10f paygManager: Update iface for ImpendingShutdown signal
✓ pick 51b2fc841 payg: Add AccountID support for the UI
✓ pick 7b025817d payg: Add AccountID info on the applet

### CODE REVIEW

File: js/misc/paygManager.js [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+    _onCodeExpired() {
+        this.emit('code-expired');
+    }
+
+    _timeRemainingSecs() {
+        if (!this._enabled)
+            return Number.MAX_SAFE_INTEGER;
+
+        return Math.max(0, this._expiryTime - (GLib.get_real_time() / GLib.USEC_PER_SEC));
```
Comment (mwleeds) [inline js/misc/paygManager.js]: This seems wrong. `g_get_real_time()` is documented to be based on wall clock time, whereas `EpgProvider:expiry-time` is based on `CLOCK_BOOTTIME`. `CLOCK_BOOTTIME` is the one we want to use; I don't think the shell should use wall clock time at all.

File: js/misc/paygManager.js [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        this._enabled = false;
+        this._expiryTime = 0;
+        this._rateLimitEndTime = 0;
+        this._codeFormat = '';
+        this._codeFormatRegex = null;
+        this._notification = null;
+
+        // Keep track of clock changes to update notifications.
+        this._wallClock = new GnomeDesktop.WallClock({ time_only: true });
+        this._wallClock.connect('notify::clock', this._clockUpdated.bind(this));
```
Comment (mwleeds) [inline js/misc/paygManager.js:99]: I think we can get rid of this. The expiration notifications give the time remaining, not a wall clock time, and we already have callbacks for when the expiry time changes.

File: js/misc/paygManager.js [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+        return Math.max(0, this._expiryTime - (GLib.get_real_time() / GLib.USEC_PER_SEC));
+    }
+
+    _clockUpdated() {
+        this._updateExpirationReminders();
+    }
+
+    _notifyPaygReminder(secondsLeft) {
+        // Only notify when in an regular session, not in GDM or initial-setup.
```
Comment (mwleeds) [inline js/misc/paygManager.js]: s/an/a/

File: js/misc/paygManager.js (hunk 1 above)
Comment (mwleeds) [reply js/misc/paygManager.js]: Now I see there's a later commit which changes it to use CLOCK_BOOTTIME. We could squash in these two:

- "paygManager: Understand ExpiryTime in terms of CLOCK_BOOTTIME"
- "payg: Use CLOCK_BOOTTIME for rate limiting"

File: js/ui/payg.js [hunk 4]
```diff
@@ -212,3 +219,249 @@ class PaygUnlockUi extends St.Widget {
...
+        // Unless we're already done, exit early so we don't destroy the
+        // notification)
+        if (this._verificationStatus !== UnlockStatus.SUCCEEDED)
+            return;
+
+        super.activate();
+    }
+});
+
+// Takes an UNIX timestamp (in seconds) and returns a string
```
Comment (mwleeds) [inline js/ui/payg.js:581]: It's really just a number of seconds not a UNIX timestamp, perhaps we should update the comment.

File: js/misc/paygManager.js [hunk 5]
```diff
@@ -174,6 +174,7 @@ var PaygManager = GObject.registerClass({
         this._expiryTime = value;
         this._updateExpirationReminders();
 
+        this._paygNotifier.clearNotification();
```
Comment (mwleeds) [inline js/misc/paygManager.js:184]: This doesn't seem to belong in this commit.

File: js/ui/payg.js [hunk 6]
```diff
@@ -356,10 +356,17 @@ var PaygUnlockWidget = GObject.registerClass({
 
 });
 
+<<<<<<< HEAD
```
Comment (mwleeds) [inline js/ui/payg.js]: merge conflicts

File: js/ui/payg.js [hunk 7]
```diff
@@ -465,3 +472,67 @@ function timeToString(seconds) {
...
+var PaygNotifier = GObject.registerClass(
+class PaygNotifier extends GObject.Object {
+    _init() {
+        super._init();
+
+        this._notification = null;
+    }
+
+    notify(secondsLeft) {
+        // Only notify when in an regular session, not in GDM or initial-setup.
```
Comment (mwleeds) [inline js/ui/payg.js:688]: There's a mismatch between the comment and the code, since we're not checking for the `gdm` or `initial-setup` modes. And similarly in `js/ui/status/automaticUpdates.js`.

File: js/ui/payg.js (hunk 6 above)
Comment (mwleeds) [reply js/ui/payg.js]: (deleted by a later commit, but still)

File: js/ui/screenShield.js [hunk 8]
```diff
@@ -155,27 +155,30 @@ var ScreenShield = class {
...
+                if (user.password_mode == AccountsService.UserPasswordMode.NONE) {
+                    this.deactivate(false);
+                    return;
+                }
+
+                let paygExpectedMode = this._isGreeter ? 'gdm-unlock-dialog-payg' : 'unlock-dialog-payg';
+                if (Main.sessionMode.currentMode === paygExpectedMode) {
+                    // This is the most common case.
+                    Main.sessionMode.popMode(paygExpectedMode);
+                }
```
Comment (mwleeds) [inline js/ui/screenShield.js:171]: What happened to the `else if (Main.sessionMode.currentMode == 'lock-screen')` which is here on the `Release_3.7.8` tag?

Comment (GeorgesStavracas) [reply js/ui/screenShield.js:171]: There is no `lock-screen` mode anymore, as of c20451c2e78ef7cb70e384cce8a7b55e37b85ce4

File: js/ui/status/payg.js [hunk 9]
```diff
@@ -29,6 +29,9 @@ class PaygIndicator extends PanelMenu.SystemIndicator {
         });
         this.menu.addMenuItem(this._item);
 
+        // Account ID
+        this._initializeAccountIDMenu();
```
Comment (mwleeds) [inline js/ui/status/payg.js:33]: Isn't this missing the changes from https://github.com/endlessm/gnome-shell/pull/632/

**Comment by mwleeds** [comment]
Finished reviewing this. Looks alright overall, thanks for rebasing it.

File: js/ui/status/payg.js (hunk 9 above)
Comment (GeorgesStavracas) [reply js/ui/status/payg.js]: It is, thanks

File: js/ui/paygUnlockDialog.js [hunk 10]
```diff
@@ -35,8 +35,6 @@ const MSEC_PER_SEC = 1000;
 // The timeout before going back automatically to the lock screen
 const IDLE_TIMEOUT_SECS = 2 * 60;
 
-const CODE_REQUIRED_LENGTH_CHARS = 8;
```
[the remark this answers is not in the document]
Comment (GeorgesStavracas) [reply js/ui/paygUnlockDialog.js]: Done

File: js/misc/paygManager.js (hunk 1 above)
Comment (GeorgesStavracas) [reply js/misc/paygManager.js]: The second commit was squashed with ba5b0ded0 but yeah, it makes sense to squash them.

File: js/misc/paygManager.js (hunk 2 above)
Comment (GeorgesStavracas) [reply js/misc/paygManager.js:99]: I'm not confident enough to drop this code and check that it works with the current limitations of time :/
